src/Toml/Type/Value.hs

```diff
@@ -0,0 +1,210 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE GADTs              #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeOperators      #-}
+
+module Toml.Type.Value
+       ( -- * Type of value
+         TValue (..)
+       , showType
+
+         -- * Value
+       , Value (..)
+       , DateTime (..)
+       , eqValueList
+       , valueType
+
+         -- * Type checking
+       , TypeMismatchError (..)
+       , sameValue
+       ) where
+
+import Data.String (IsString (..))
+import Data.Text (Text)
+import Data.Time (Day, LocalTime, TimeOfDay, ZonedTime, zonedTimeToUTC)
+import Data.Type.Equality ((:~:) (..))
+
+-- | Needed for GADT parameterization
+data TValue = TBool | TInteger | TDouble | TText | TDate | TArray
+    deriving (Eq, Show)
+
+showType :: TValue -> String
+showType = drop 1 . show
+
+-- TODO: examples are copy-pasted from TOML specification. Probably most of them
+-- will be moved into parsing module in future.
+-- | Value in @key = value@ pair.
+data Value (t :: TValue) where
+    {- | Boolean value:
+
+@
+bool1 = true
+bool2 = false
+@
+    -}
+    Bool :: Bool -> Value 'TBool
+
+    {- | Integer value:
+
+@
+int1 = +99
+int2 = 42
+int3 = 0
+int4 = -17
+int5 = 5_349_221
+hex1 = 0xDEADBEEF
+oct2 = 0o755 # useful for Unix file permissions
+bin1 = 0b11010110
+@
+    -}
+    Integer :: Integer -> Value 'TInteger
+
+    {- | Floating point number:
+
+@
+flt1 = -3.1415   # fractional
+flt2 = 1e6       # exponent
+flt3 = 6.626e-34 # both
+flt4 = 9_224_617.445_991_228_313
+@
+    -}
+    Double :: Double -> Value 'TDouble
+
+    {- | String value:
+
+@
+key = "value"
+bare_key = "value"
+bare-key = "value"
+@
+    -}
+    Text :: Text -> Value 'TText
+
+    -- | Date-time. See documentation for 'DateTime' type.
+    Date :: DateTime -> Value 'TDate
+
+    {- | Array of values. According to TOML specification all values in array
+      should have the same type. This is guaranteed statically with this type.
+
+@
+arr1 = [ 1, 2, 3 ]
+arr2 = [ "red", "yellow", "green" ]
+arr3 = [ [ 1, 2 ], [3, 4, 5] ]
+arr4 = [ "all", 'strings', """are the same""", '''type''']
+arr5 = [ [ 1, 2 ], ["a", "b", "c"] ]
+
+arr6 = [ 1, 2.0 ] # INVALID
+@
+    -}
+    Array  :: [Value t] -> Value 'TArray
```

Comment:
Since this is implemented as a Haskell list, should it be changed to `List` instead of `Array`?

Reply:
I would like to keep it as `Array` by the following reasons:

1. In Haskell types like `Int`, `String` and `Float` are very common and well-known and are different from `Integer`, `Text` and `Double` (which are also well-known Haskell types). This leads to confusion in using data types. But there's no such type like `Array` (at least commonly-used). So it's okay to be closer to toml specification here.
2. In future internal representation of `Array` constructor might change to something like `Vector` if we discover some serious performance problems. So it no longer will be `List` and having `Array` will help with backwards compatibility. 
3. `Text`, `Integer` and `Double` in Haskell are semantically exactly equivalent to toml specification. While lists are not semantically equivalent. In toml you can have an array like `[ [1, 2], [true, false] ]` but in Haskell you can't. So having `Array` name for variable hints user that it's different from Haskell lists.